### ISSUE TITLE
Allow subdirectories within git repos in yarn install

### ISSUE THREAD

**Issue by akshetpandey** [yarnpkg/yarn#4725]
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
Unable to install packages that are inside subdirectories in git repos. This is a very common case for monorepos and just repositories that can't have the node package in the root repo.
 
**If the current behavior is a bug, please provide the steps to reproduce.**
Not a but

**What is the expected behavior?**
There should be an availably syntax for `yarn add` and `yarn global add` that allows to add packages that are inside subdirectories in a git repo.

It make it much easier to use unreleased code when testing unreleased versions for issues/regressions or fixes that aren't live yet like this ticket https://github.com/Microsoft/code-push/issues/428

ex. packages affected by this issue
https://github.com/Microsoft/code-push/tree/master/cli
https://github.com/Microsoft/BotBuilder/tree/master/Node
https://github.com/facebook/react-native/tree/master/react-native-cli
https://github.com/babel/babel/tree/master/packages/babel-cli

Related npm issue: https://github.com/npm/npm/issues/2974

**Please mention your node.js, yarn and operating system version.**
yarn versions v1.2.1
node version v8.7.0
macOS Sierra 10.12.6

**Comment by akshetpandey** [comment]
There are a couple of good ideas for syntax to use to specify subdirectories in the linked issue (npm/npm#2974)

Use pip's way of using a query parameter
`git+https://git.repo/some_repo.git?subdirectory=subdir_path#branch`
or
`git+https://git.repo/some_repo.git#branch?path=subdir_path`
Although I would assume most url parsers won't parse this one.


Use `::` as it is not a valid token for a branch name
`git+https://git.repo/some_repo.git#branch::subdir_path`
`git+https://git.repo/some_repo.git#::subdir_path`

**Comment by lukescott** [comment]
How about `git+https://git.repo/some_repo.git/subdir_path#branch`.

**Comment by akshetpandey** [comment]
That might be ambiguous depending on how the git server urls are setup.

The url might be something like

`git+https://github.com/yarnpkg/yarn/<subdir_path>#branch`

Then it become hard to say what is part of the repo path and what is the subdir_path

**Comment by lukescott** [comment]
Not sure if this applies to every repo, but with Github both `git+https://github.com/yarnpkg/yarn` and `git+https://github.com/yarnpkg/yarn.git` work. You could assume anything after `.git/` is a sub-directory.

Regardless of syntax it would be nice to have this feature either way. It would nice to make the syntax intuitive though :).

**Comment by akshetpandey** [comment]
Oh yea, I was mostly talking about other website, not really github. Sorry my example wasn't the best one to explain it.

**Comment by stealthybox** [comment]
Long discussion about supporting this in npm:
https://github.com/npm/npm/issues/2974

There are some suggested separators in the thread.

**Comment by joshcartme** [comment]
I would also like to be able to do this to facilitate working with a monorepo that contains a lot of sub modules.  I also found #1570 which discusses this.  That was closed in favor of https://github.com/yarnpkg/rfcs/issues/15 but #15 doesn't exist anymore so it's hard to tell what if anything was done.

**Comment by matheo** [comment]
so `yarn` doesn't support this yet?
all right, no simpler option than publish and use npm packages :(

**Comment by erykpiast** [comment]
Yeah, rising popularity of monorepo architectures makes it a crucial feature. Is there any workaround for this, for now? Discussion about public API tend to be years long...

**Comment by lukescott** [comment]
My #1 use-case for this is pulling in bug-fixes from monorepo's that haven't been merged or released yet. I often want to keep most of the original packages, but pull in one of the sub-repos with a bug-fix. Currently I am unable to do this, and the workarounds are complex. Being able to reference a Github branch would greatly lower the resistance.

I know there are some people at NPM that are highly resistant to the idea of this, and monorepos in general. However, many of the contributors here work on projects that use monorepos. I would suspect that if yarn were to implement this feature and it became popular enough npm would follow suit. A lot of yarn ideas have been implemented in npm later on. yarn to be is the babel of npm.

**Comment by sdhull** [comment]
@erykpiast couple workarounds that I can think of:
1. clone the target repo somewhere and then use `yarn add /path/to/repo/subdirectory`
2. or possibly (if it's in your own monorepo) use yarn [workspaces](https://yarnpkg.com/lang/en/docs/workspaces/)?

Or -- this is maybe even crazier -- add the remote in question as a [submodule](https://github.com/blog/2104-working-with-submodules) and (optionally move your project files into a subdirectory) _then_ use yarn workspaces?

Disclaimer: I haven't tried either of these approaches and they both add overhead for new developers & your CI builds...

**Comment by jwdunne** [comment]
> Or -- this is maybe even crazier -- add the remote in question as a submodule and (optionally move your project files into a subdirectory) then use yarn workspaces?

I can confirm this works.

Upgraded to webpack 4. Create React App doesn't have support yet but there's a PR that's waiting to be merged - unfortunately webpack 4 has changed it's plugin API. Doubly unfortunate: create-react-app is repo containing multiple packages. Adding the forked PR as a submodule and adding the package using 'file:' syntax worked a charm. 

Literally saved an entire day or more of work with this work around.

**Comment by micimize** [comment]
It's alarming that this isn't on the roadmap for either npm or yarn due to the rise in popularity of monorepos. The `git+...` syntax makes contributing to single-package projects a breeze, and by contrast not having it for subdirectories makes contributing to monorepo projects a total headache.

**Comment by jwdunne** [comment]
I agree. After a few months down the git submodule, `file:` route, it's not ideal. Onboarding other developers is more difficult. My colleagues have to worry about initialising and updating the submodules. In their defence, it used to be as simple as running `yarn` to get themselves going.

Monorepo support would be a dream come true.

**Comment by Frikki** [comment]
yarn + lerna (using yarn’s workspaces) works to manage monorepos.

**Comment by micimize** [comment]
@Frikki I don't believe lerna has much [submodule support](https://github.com/lerna/lerna/issues/161), and isn't a "consumption tool" anyways.

@jwdunne For now I just `npm publish` a namespaced version of the fork and depend on it

**Comment by ramasilveyra** [comment]
Hello! I built https://github.com/ramasilveyra/gitpkg to "publish" packages from monorepos as git tags, it may help.

**Comment by stealthybox** [comment]
@ramasilveyra that looks super useful -- thanks for the addition 👍

**Comment by robermorales** [comment]
Some news in supporting git subdirs as install sources?

Thanks,

**Comment by nickiaconis** [comment]
I encountered this issue today when needing to `yarn add` a yet unaccepted/unpublished fix I wrote for a monorepo. +1 from me.

**Comment by UrielCh** [comment]
Foking [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) cause the same king of issue,
Each time I fork this git, to prepare a new pull request, I need to depended on a sub directory of my forked git.
Is `file://` is currently the only solution ?

**Comment by nickiaconis** [comment]
@UrielCh I was able to use https://github.com/ramasilveyra/gitpkg as a workaround. It might work for your situation too.

**Comment by UrielCh** [comment]
Thx for this suggestion, creating a new git for this fork will be simple.

**Comment by t-benze** [comment]
> Hello! I built https://github.com/ramasilveyra/gitpkg to "publish" packages from monorepos as git tags, it may help.

This may look a bit hacky at the beginning, but after some reasoning it seems to be a legitimate solution to the problem 👍

**Comment by yanickrochon** [comment]
Seriously, [gitpkg](https://github.com/ramasilveyra/gitpkg) works! It creates tags inside the project repository, and can be used with repositories containing multiple packages.

And this prompted me to ditch `npm` because only `yarn` can fetch nicely from Git.

**Comment by Preen** [comment]
+1

**Comment by micimize** [comment]
> Hello! I built https://github.com/ramasilveyra/gitpkg to "publish" packages from monorepos as git tags, it may help.

Just used `gitpkg` and I think it's such a good solution that it should be added as a note in the docs for this use case

**Comment by EqualMa** [comment]
I built a service https://gitpkg.now.sh/ to download the sub folder of a git repo as a `.tgz` file, so that it can be installed by yarn and npm. With this service, package users can use the subdir directly and there is no need for package developers to publish the subdirs.

For example, if you want to add [`babel-core` in branch `next-8-dev`](https://github.com/babel/babel/tree/next-8-dev/packages/babel-core) as dependency,
just use:

```
yarn add https://gitpkg.now.sh/babel/babel/packages/babel-core?next-8-dev
```

also works for `npm install`.

Note that I must say sorry to @ramasilveyra  because my service's name conflicts with the [tool](https://github.com/ramasilveyra/gitpkg) he develops but I didn't know that before.

**Comment by dword-design** [comment]
@EqualMa This looks pretty amazing. I'll have a look at this soo.

**Comment by erykpiast** [comment]
@EqualMa It'd be even more amazing if it works for projects with a build step. But anyway, great job, thank you so much!

**Comment by uriel-sf** [comment]
Do we still need gitpkg? Running latest lerna and when run in independent-mode it will publish git tags of each component.

**Comment by EqualMa** [comment]
@erykpiast Finally [GitPkg service](https://gitpkg.now.sh) supports custom scripts, allowing you to run a build step in postinstall or other scripts! You can checkout the [docs](https://gitpkg.now.sh/guide/#custom-scripts) or just try adding custom scripts at https://gitpkg.now.sh

**Comment by erykpiast** [comment]
@EqualMa that's even more than awesome! Thank you!

**Comment by Acidic9** [comment]
Any updates to this? Is it planned or anything?

**Comment by artemjackson** [comment]
@EqualMa awesome! great job!

**Comment by jason-krypton** [comment]
I'm having an issue trying to fork a repo (https://github.com/ianstormtaylor/slate) that uses workspaces and has multiple packages in the same repo. I only want to modify and import one of the packages. Trying to `yarn add` pointing to the directory that holds the individual package.json file gives an error...

```
yarn add https://github.com/kr-project/slate/tree/master/packages/slate#e73db650a332fc5ded3fd72cffb1fcd7279b85d7
error https://github.com/kr-project/slate/tree/master/packages/slate: Extracting tar content of undefined failed, the file appears to be corrupt: "Invalid tar header. Maybe the tar is corrupted or it needs to be gunzipped?"
```

I've also tried to check in the published files, but my understanding is this is a problem with yarn looking a tar only found at the root directory of a github repo?

So is the only way to fork a project that uses workspaces to use something like `gitpkg`? I'm a bit surprised that there doesn't seem to be more native support for this.

**Comment by LucidDan** [comment]
Just ran into exactly the same issue as @jason-krypton ... trying to work with a fork of a repo that uses workspaces, normally I'd just yarn add the forked git repository ... that's impossible now I guess? Wasting a decent bunch of time on this in order to change a couple of lines of code. :-(

**Comment by robclancy** [comment]
I have similar issues with ckeditor5. I'm trying out GitPkg now.

**Comment by bedney** [comment]
So @LucidDan and others.

I've found the following to work:

- You *must* be running `yarn 2.1.X` for this to work.
- See this page for the newer syntax that allows this: https://yarnpkg.com/features/protocols
- You have to add individual entries *for each package in the monorepo* to your `package.json` under either `dependencies` or, in my case, `devDependencies`:

```
"package-1": "my-org-name/mono-repo-name#head=my-git-branch&workspace=package-1"
```

It's disappointing that there isn't a way to just have `yarn` install the top-level monorepo and unroll all of the attendant packages into their proper place. Maybe there is and I'm just being stupid here? If so, someone enlighten me ;-)

**Comment by paul-soporan** [comment]
Implemented in v2, starting from Yarn 2.1: https://dev.to/arcanis/yarn-2-1-git-workspaces-focused-installs-loose-mode-live-playground-4kfc#workspace-cloning

You can use the following syntax to install a specific workspace (not arbitrary folder):
```
yarn add @yarnpkg/eslint-config@yarnpkg/berry#workspace=@yarnpkg/eslint-config
```

https://yarnpkg.com/getting-started/migration

**Comment by nomi9995** [comment]
Try https://gitpkg.now.sh/

```shell
yarn add https://gitpkg.now.sh/<user>/<repo>/<subdir>[?<branch>]
```
OR
```
"@react-native-community/cli-platform-ios":"https://gitpkg.now.sh/nomi9995/cli/packages/platform-ios?fix-ios-getProductName-regex"
```
